lightning-block-sync/src/http_endpoint.rs

```diff
@@ -0,0 +1,178 @@
+use http;
```

Comment:
> And as we move outward to crates like lightning-block-sync we may want to allow for some others, possibly only in sample implementations they provide like the HTTP clients.

Totally agreed here.

> My argument for using this particular crate is that it is fairly lightweight and lets us focus on domain-specific code.

Yea, if it were more lightweight I may agree, but I think that's the wrong metric - I think the right one would be something that captures the differing costs - how many LoC and of what complexity would it take to write our own, and how many extra dependencies are involved (here its, I believe, three crates just to take http, but it may be more). For JSON, the dependency tree is steep, but JSON is also rather difficult to parse correctly, and easy to screw up. For URI parsing, we're just looking for some separators and returning a few parts, it should be rather difficult to screw that up. As for `read_http_resp`, I'd *love* to replace that with a "real" HTTP crate, but last I heard all the options were downright terrible (https://medium.com/@shnatsel/smoke-testing-rust-http-clients-b8f2ee5db4e6 has been updated to indicate its rather out-of-date, but I haven't done any research on what the latest state is).

Reply:
> Yea, if it were more lightweight I may agree, but I think that's the wrong metric - I think the right one would be something that captures the differing costs - how many LoC and of what complexity would it take to write our own, and how many extra dependencies are involved (here its, I believe, three crates just to take http, but it may be more). For JSON, the dependency tree is steep, but JSON is also rather difficult to parse correctly, and easy to screw up. For URI parsing, we're just looking for some separators and returning a few parts, it should be rather difficult to screw that up. As for `read_http_resp`, I'd _love_ to replace that with a "real" HTTP crate, but last I heard all the options were downright terrible (https://medium.com/@shnatsel/smoke-testing-rust-http-clients-b8f2ee5db4e6 has been updated to indicate its rather out-of-date, but I haven't done any research on what the latest state is).

Yeah, it takes on three dependencies:

```
├── http v0.2.1
│   ├── bytes v0.5.5
│   ├── fnv v1.0.7
│   └── itoa v0.4.6
```

For URI parsing, the implementation that I replaced was not compatible with Rust's `std::net::ToSocketAddrs` for some cases. So there's an argument in how much time we want to invest in making sure the implementation is robust and well tested vs grabbing something off the shelf.

Reply:
As for `read_http_resp`, https://crates.io/crates/chunked_transfer ! Has zero dependencies, too.

Reply:
Will that crate work with `tokio`? It requires sources to implement `std::io::Read`.

Reply:
This one seems more suitable for async stream processing: https://docs.rs/uhttp_chunked_bytes/0.5.0/uhttp_chunked_bytes/ - it can defer when data is not yet available.

Reply:
Nice find! I'll take a stab at using it.

Reply:
> This one seems more suitable for async stream processing: https://docs.rs/uhttp_chunked_bytes/0.5.0/uhttp_chunked_bytes/ - it can defer when data is not yet available.

Turns out `uhttp_chunked_bytes` requires an `Iterator` over the body, which is problematic because iterators cannot be async. We'd have to read the entire stream into memory (assuming the server disconnects) then pass an iterator over the read buffer to the library. Let me know if you had something else in mind.

The same approach can be used with `chunked_transfer`, which has an easier interface to work with, IMHO. Need to add some DoS protection, but I think this should be doable using `AsyncReadExt::take`.

Reply:
You are right about `uhttp_chunked_bytes` - I misread the code - it doesn't seem like it can defer in the middle of a read.

About reading the body - isn't the issue that with chunked encoding you don't know how long the body is, so you can't read it fully without doing the decoding as you go along?

Reply:
Right, the approach only works if the server closes the stream, as then the client would get an EOF. Not sure if this is a safe assumption though. Otherwise, the only way of knowing if you hit the end is by doing the decoding, as you said.

Reply:
I ended up writing a simple adapter using `futures::executor::block_on`.

```rust
struct ReadAdapter<'a, R: tokio::io::AsyncRead + std::marker::Unpin>(&'a mut R);

impl<'a, R: tokio::io::AsyncRead + std::marker::Unpin> std::io::Read for ReadAdapter<'a, R> {
       fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
               #[cfg(test)]
               std::thread::yield_now();
               futures::executor::block_on(self.0.read(buf))
       }
}
```

It does the trick, but in order to consistently avoid test deadlocks I had to have the `HttpServer` drop its side of the `TcpStream` after writing the response.